torch/csrc/autograd/input_buffer.cpp

```diff
@@ -129,96 +149,147 @@ static void accumulate(
+
+  //
+  // [ Non-accelerator case ]
+  //
+  if (!is_accelerator) {
+    if (current_index == 0) {
+      TORCH_INTERNAL_ASSERT(!buffer[pos].defined())
+      buffer[pos] = std::move(var);
     } else {
-      std::optional<c10::Stream> opt_sync_stream = std::nullopt;
-      const auto guard = c10::impl::VirtualGuardImpl{device_type};
-      if (on_consumer && !on_producer) {
-        // (3a)
-        opt_accumulate_stream = opt_consumer_stream;
-        opt_sync_stream = guard.getDefaultStream(opt_consumer_stream->device());
-      } else if (on_producer && !on_consumer) {
-        // (4a)
-        opt_accumulate_stream =
-            guard.getDefaultStream(opt_producer_stream->device());
-        opt_sync_stream = opt_producer_stream;
-      } else {
-        // (5)
-        opt_accumulate_stream = guard.getDefaultStream(*device);
-      }
-      if (opt_sync_stream && (opt_accumulate_stream != opt_sync_stream)) {
-        // (3b), (4b)
-        c10::OptionalDeviceGuard device_guard{opt_sync_stream->device()};
-        auto event = c10::Event{device_type};
-        event.record(*opt_sync_stream);
-        opt_accumulate_stream->wait(event);
-        const auto guard = c10::impl::VirtualGuardImpl(device_type);
-        record_stream_any_impl(var, *opt_accumulate_stream);
-      }
+      c10::OptionalDeviceGuard device_guard{device};
+      accumulate(buffer, pos, std::move(var));
     }
+    return;
   }
 
-  auto& old_var = buffer[pos];
-  if (!old_var.defined()) {
+  // Handle the case where var is on accelerator but producer node has no
+  // canonical stream, e.g. this can happen if forward is DtoH
+  const std::optional<c10::Stream>& opt_producer_stream =
+      (opt_producer_stream_.has_value()
+           ? opt_producer_stream_
+           : std::optional<c10::Stream>(
+                 at::accelerator::getCurrentStream(device->index())));
+
+  TORCH_INTERNAL_ASSERT(opt_consumer_stream && opt_producer_stream);
+
+  //
+  // [ Single producer case ]
+  //
+  if (num_dependencies == 1) {
+    mb_wait_stream(opt_consumer_stream, opt_producer_stream, device_type);
+    mb_record_stream_any_impl(
+        var, *opt_consumer_stream, /*prev_stream=*/*opt_producer_stream);
+    TORCH_INTERNAL_ASSERT(current_index == 0);
+    TORCH_INTERNAL_ASSERT(!buffer[pos].defined());
     buffer[pos] = std::move(var);
-  } else {
-    if (opt_accumulate_stream) {
-      c10::OptionalStreamGuard stream_guard{opt_accumulate_stream};
-      accumulate(buffer, pos, std::move(var));
+    return;
+  }
+  //
+  // [ Multiple producer case ]
+  //
+  // First producer
+  // ==============
+  //
+  // A. Determine the accumulation stream:
+  //    case 1) var's device matches consumer node's canonical device
+  //            (The producer node's canonical device may or may not match)
+  //            -> accumulator stream = consumer stream
+  //    case 2) var's device only matches producer node's canonical device
```

Comment:
Not sure what this "only" means here.
You mean it doesn't match the consumer but does match the producer?

Reply:
Yeah, will update comment to clarify